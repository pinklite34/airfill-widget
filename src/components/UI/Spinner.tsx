import React from 'react';
import PropTypes from 'prop-types';
import styled from 'react-emotion';

const duration = '0.7s';

const Container = styled('div')`
  display: flex;
  width: 100%;
  height: 100%;
  min-height: ${p => (p.tight ? 0 : '100px')};
  align-items: center;
  justify-content: center;

  animation: ${p => p.fadeIn && `fadein ${p.fadeIn}s`};

  @keyframes fadein {
    0% {
      opacity: 0;
    }
    25% {
      opacity: 0;
    }
    75% {
      opacity: 0.5;
    }
    100% {
      opacity: 1;
    }
  }
`;

const Outer = styled('div')`
  @keyframes anim-spinner {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }

  display: inline-block;
  animation-name: anim-spinner;
  animation-duration: ${duration};
  animation-iteration-count: infinite;
  animation-timing-function: linear;

  * {
    box-sizing: border-box;
  }
`;

const Circle1 = styled('div')`
  width: ${p => p.size}em;
  height: ${p => p.size / 2}em;
  overflow: hidden;
`;

const Circle2 = styled('div')`
  width: ${p => p.size}em;
  height: ${p => p.size / 2}em;
  overflow: hidden;
  transform: rotate(180deg);
`;

const Inner = styled('div')`
  @keyframes anim-circle-1 {
    from {
      transform: rotate(60deg);
    }
    to {
      transform: rotate(205deg);
    }
  }

  @keyframes anim-circle-2 {
    from {
      transform: rotate(30deg);
    }
    to {
      transform: rotate(-115deg);
    }
  }

  transform: rotate(45deg);
  border-radius: 50%;
  border: ${p => p.size / 8}em solid
    ${p => (p.white ? p.theme.white : p.theme.brand)};
  border-right: ${p => p.size / 8}em solid transparent;
  border-bottom: ${p => p.size / 8}em solid transparent;
  width: 100%;
  height: 200%;
  animation-name: ${p => (p.second ? 'anim-circle-2' : 'anim-circle-1')};
  animation-duration: ${duration};
  animation-iteration-count: infinite;
  animation-direction: alternate;
  animation-timing-function: cubic-bezier(0.25, 0.1, 0.5, 1);
`;

export default function Spinner({ small, ...props }) {
  const size = small ? 1 : 2;
  return (
    <Container size={size} {...props}>
      <Outer size={size} {...props}>
        <Circle1 size={size} {...props}>
          <Inner size={size} {...props} />
        </Circle1>
        <Circle2 size={size} {...props}>
          <Inner second size={size} {...props} />
        </Circle2>
      </Outer>
    </Container>
  );
}

Spinner.propTypes = {
  tight: PropTypes.bool,
  white: PropTypes.bool,
  small: PropTypes.bool,
  fadeIn: PropTypes.number,
};

Spinner.defaultProps = {
  fadeIn: 1,
};
