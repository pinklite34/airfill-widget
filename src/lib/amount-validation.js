import { getDisplayName } from './currency-helpers';

// Handle XBT
export const getCostForAmount = (conversionRate, amount) =>
  Number(amount * conversionRate).toFixed(2);

const pickMiddlePackage = ({ packages }) => {
  const middle = Math.round((packages.length - 1) * 0.6);
  const pkg = packages[middle];
  return pkg ? String(pkg.value) : null;
};

const pickAffordablePackage = ({ packages, amount, currency, maxCost }) => {
  const costKey = getDisplayName(currency).toLowerCase() + 'Price';

  const isAffordable = pkg => pkg[costKey] <= maxCost;

  // Make sure user can afford the selected package
  const selectedPackage = packages.find(pkg => String(pkg.value) === amount);
  if (selectedPackage && isAffordable(selectedPackage)) {
    return amount;
  }

  // Otherwise try to pick the highest value package the user can afford
  const highestAffordablePackage = packages.filter(isAffordable).pop();

  if (highestAffordablePackage) {
    return Number(highestAffordablePackage.value);
  }

  // If there are no packages the user can afford, return the amount as is
  return amount;
};

export const selectValidPackage = ({ packages, maxCost, currency, amount }) => {
  // If no amount is selected, pick a package in the middle
  amount = amount ? String(amount) : pickMiddlePackage({ packages });

  // Make sure user can afford the package
  return pickAffordablePackage({ packages, amount, maxCost, currency });
};

const rangedCostForAmount = (conversionRate, currency, amount) => {
  let amountCost = amount * conversionRate;
  // BTC
  if (currency === 'XBT') {
    return Math.ceil(amountCost / 100) / 1000000;
  } else {
    // USD or EUR
    return Number(amountCost.toFixed(2));
  }
};

const selectValidRangedAmount = ({
  amount,
  maxCost,
  costConversionRate,
  currency
}) => {
  const selectedAmountCost = rangedCostForAmount(
    costConversionRate,
    currency,
    amount
  );

  if (selectedAmountCost <= maxCost) {
    return String(amount); // Return amount as is for ranged operators
  } else {
    if (currency === 'XBT') {
      const amountForMaxCost = Math.floor(
        maxCost * 100000000 / costConversionRate
      );
      return String(amountForMaxCost); // Return the maximum amount allowed
    } else {
      return String(maxCost / costConversionRate);
    }
  }
};

export const selectValidAmount = args => {
  const ranged = args.ranged;
  const validPackageAmount = selectValidPackage(args);

  if (ranged) {
    return selectValidRangedAmount({
      ...args,
      amount: args.amount || validPackageAmount
    });
  } else {
    return validPackageAmount;
  }
};

