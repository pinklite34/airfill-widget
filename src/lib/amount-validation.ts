import { Amount, BillingCurrency, Package, RangeProp } from '../types';

import { getDisplayName } from './currency-helpers';

// if package is affordable with current user balance
// is affordable if no balance is set
export function isAffordable(
  pkg: Package,
  billingCurrency: BillingCurrency,
  balance: Amount
) {
  const costKey = getDisplayName(billingCurrency).toLowerCase() + 'Price';

  return !balance || pkg[costKey] <= balance;
}

export function selectValidAmount(
  packages: Package[],
  billingCurrency: BillingCurrency,
  balance?: Amount,
  range?: RangeProp
) {
  if (range) {
    const correctMax =
      billingCurrency === 'XBT'
        ? ((balance as number) * 100000000) / range.userPriceRate
        : (balance as number) / range.userPriceRate;

    if (correctMax < range.min) {
      return 0;
    } else if (correctMax < range.max) {
      return correctMax.toFixed();
    } else {
      return range.max;
    }
  } else {
    const costKey = getDisplayName(billingCurrency).toLowerCase() + 'Price';

    const f = packages
      .sort((a, b) => (a[costKey] < b[costKey] ? 1 : -1))
      .filter(pkg => isAffordable(pkg, billingCurrency, balance));

    // pick highest affordable or most expensive, if no affordable
    return f.length > 0 ? f[0].value : packages[0].value;
  }
}
