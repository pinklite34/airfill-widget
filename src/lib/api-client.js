import ponyFetch from 'fetch-ponyfill';
const { fetch: _fetch } = ponyFetch();

export function encodeQueryString(obj) {
  const queryString =
    typeof obj === 'object' &&
    !!obj &&
    Object.keys(obj)
      .filter(k => obj[k])
      .map(k => encodeURIComponent(k) + '=' + encodeURIComponent(obj[k]))
      .join('&');

  return queryString ? '?' + queryString : '';
}

export const createClient = (conf = {}) => {
  let baseUrl = conf.baseUrl || '/api';
  let { username, password, defaultHeaders, token } = conf;

  const doFetch = (uri, options) => {
    let url = baseUrl + uri;

    // Set default headers
    options.headers = {
      Accept: 'application/json',
      ...defaultHeaders,
      ...options.headers,
    };

    // Build query string if needed
    if (options.query) {
      url = url + encodeQueryString(options.query);
    }

    // Add body data
    if (options.body) {
      options.headers = {
        'Content-Type': 'application/json',
        ...options.headers,
      };

      if (typeof options.body !== 'string') {
        options.body = JSON.stringify(options.body);
      }
    }

    options.method = options.method || (options.body ? 'post' : 'get');

    let user = options.username || username;
    let pass = options.password || password;

    // Add auth if provided
    if (user && pass) {
      options.headers = {
        Authorization: 'Basic ' + btoa(user + ':' + pass),
        ...options.headers,
      };
    } else if (token) {
      options.headers = {
        Authorization: 'Token ' + token,
        ...options.headers,
      };
    }

    return _fetch(url, options)
      .then(response => {
        if (response.status >= 200 && response.status < 300) {
          return response.json();
        }
        throw response;
      })
      .catch(error => {
        if (typeof error && error.json === 'function') {
          return error
            .json()
            .then(json => {
              error.json = json;

              if (json.error) {
                return json.error.message;
              }
              if (json.message) {
                return json.message;
              }

              throw Error('No error information in JSON response');
            })
            .catch(() => {
              if (error.statusText) {
                return error.statusText;
              } else {
                return 'Server returned status code ' + error.status;
              }
            })
            .then(message => {
              const err = new Error(message);
              err.response = error;
              throw err;
            });
        } else {
          console.error(error.message || error);
        }
      });
  };

  const configure = conf => {
    baseUrl = 'baseUrl' in conf ? conf.baseUrl : baseUrl;
    username = 'username' in conf ? conf.username : username;
    password = 'password' in conf ? conf.password : password;
    token = 'token' in conf ? conf.token : token;
    defaultHeaders =
      'defaultHeaders' in conf ? conf.defaultHeaders : defaultHeaders;
  };

  const createHref = (uri, queryParams) => {
    return baseUrl + (uri || '') + encodeQueryString(queryParams);
  };

  return { fetch: doFetch, configure, createHref };
};

export const client = createClient();

export function fetch(uri, options) {
  return client.fetch(uri, options);
}
