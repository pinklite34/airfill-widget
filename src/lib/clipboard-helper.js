export default text => {
  const element = document.createElement('textarea');
  element.textContent = text;
  document.body.appendChild(element);
  element.select();

  try {
    document.execCommand('copy');
  } catch (e) {}

  document.body.removeChild(element);
  window.alert('Copied to clipboard!');
};
