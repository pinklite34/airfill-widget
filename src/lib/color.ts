import * as Color from 'color';

export function colorToString(color) {
  return Color(color)
    .hsl()
    .string();
}

export function colorDarken(color, amount) {
  return colorToString(Color(color).darken(amount));
}

export function colorLighten(color, amount) {
  return colorToString(Color(color).lighten(amount));
}
