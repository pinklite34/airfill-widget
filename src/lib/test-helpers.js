import renderer from 'react-test-renderer';

export const matchRenderWithSnapshot = jsx => {
  const tree = renderer.create(jsx).toJSON();
  expect(tree).toMatchSnapshot();
}
