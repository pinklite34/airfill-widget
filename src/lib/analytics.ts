import { Order, Product, RootState } from '../types';

import { getPlatform, getSource, isMobileApp } from '../lib/globals';

export function productPropertiesForOrder(order?: Order) {
  return order
    ? {
        product_id: order.operator,
        sku: order.operator,
        name: order.itemDesc,
        price: order.eurPrice,
        quantity: 1,
        category: order.operatorType,
        currency: 'EUR',
      }
    : {};
}

export function eventPropertiesForOrder(order?: Order) {
  return order
    ? {
        ...order,
        order_id: order.id,
        affiliation: 'bitrefill',
        value: order.eurPrice,
        revenue: order.eurPrice,
        label: order.paymentMethod,
        currency: 'EUR',
        products: [productPropertiesForOrder(order)],
      }
    : null;
}

export function eventPropertiesForProduct(product?: Product) {
  return product
    ? {
        ...product,
        product_id: product.slug,
        sku: product.slug,
        category: product.type,
        name,
        image_url: product.logoImage,
      }
    : null;
}

export function createEvent(
  state: RootState,
  eventType: string,
  payload
): object {
  payload = payload || { properties: {} };

  const source = getSource();
  const platform = getPlatform();

  const additionalKey = payload.traits ? 'traits' : 'properties';
  const additionalProps = {
    category: isMobileApp() ? platform : 'website',
    user_source: source,
    user_source_platform: platform,
  };

  return {
    eventType,
    eventPayload: {
      ...payload,
      [additionalKey]: {
        ...additionalProps,
        ...payload[additionalKey],
      },
    },
  };
}
