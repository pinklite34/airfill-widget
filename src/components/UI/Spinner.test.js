import React from 'react';
import { matchRenderWithSnapshot } from '../../lib/test-helpers';

import Spinner from './Spinner';

describe('Spinner', () => {
  it('renders correctly', () => {
    matchRenderWithSnapshot(<Spinner />);
  });

  it('renders with custom text', () => {
    matchRenderWithSnapshot(<Spinner>Custom loading text</Spinner>);
  });

  it('renders without text', () => {
    matchRenderWithSnapshot(<Spinner hideText />);
  });

  it('renders with inverted animation', () => {
    matchRenderWithSnapshot(<Spinner inverted />);
  });

  it('renders inline', () => {
    matchRenderWithSnapshot(<Spinner inline />);
  });
});
