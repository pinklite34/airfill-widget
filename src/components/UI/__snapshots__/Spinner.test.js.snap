// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner renders correctly 1`] = `
<div
  className="gLiHgo"
>
  <div
    className="gcAxYX"
  >
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
  </div>
  <div
    className="kHkHSj"
  >
    Loading...
  </div>
</div>
`;

exports[`Spinner renders inline 1`] = `
<div
  className="cnXlTn"
>
  <div
    className="gcAxYX"
  >
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
  </div>
  <div
    className="kHkHSj"
  >
    Loading...
  </div>
</div>
`;

exports[`Spinner renders with custom text 1`] = `
<div
  className="gLiHgo"
>
  <div
    className="gcAxYX"
  >
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
  </div>
  <div
    className="kHkHSj"
  >
    Custom loading text
  </div>
</div>
`;

exports[`Spinner renders with inverted animation 1`] = `
<div
  className="gLiHgo"
>
  <div
    className="gcAxYX"
  >
    <div
      className="fyPcrT"
    />
    <div
      className="fyPcrT"
    />
    <div
      className="fyPcrT"
    />
    <div
      className="fyPcrT"
    />
    <div
      className="fyPcrT"
    />
    <div
      className="fyPcrT"
    />
    <div
      className="fyPcrT"
    />
    <div
      className="fyPcrT"
    />
    <div
      className="fyPcrT"
    />
    <div
      className="fyPcrT"
    />
    <div
      className="fyPcrT"
    />
    <div
      className="fyPcrT"
    />
  </div>
  <div
    className="kHkHSj"
  >
    Loading...
  </div>
</div>
`;

exports[`Spinner renders without text 1`] = `
<div
  className="gLiHgo"
>
  <div
    className="gcAxYX"
  >
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
    <div
      className="imwLvD"
    />
  </div>
</div>
`;
